packages/keystone/databaseAdapters/BalancingReplicaKnexAdapter/adapter.js

```diff
@@ -0,0 +1,193 @@
+const { KnexAdapter } = require('@keystonejs/adapter-knex')
+const { versionGreaterOrEqualTo } = require('@keystonejs/utils')
+const get = require('lodash/get')
+const omit = require('lodash/omit')
+
+const conf = require('@open-condo/config')
+const { _internalGetAsyncLocalStorage } = require('@open-condo/keystone/executionContext')
+
+const { KnexPool } = require('./pool')
+const { getNamedDBs, getReplicaPoolsConfig, getQueryRoutingRules, isDefaultRule } = require('./utils/env')
+const { initKnexClient } = require('./utils/knex')
+const { logger } = require('./utils/logger')
+const { isRuleMatching } = require('./utils/rules')
+const { extractCRUDQueryData } = require('./utils/sql')
+
+
+const graphqlCtx = _internalGetAsyncLocalStorage('graphqlCtx')
+
+class BalancingReplicaKnexAdapter extends KnexAdapter {
+    constructor ({ databaseUrl, replicaPools, routingRules }) {
+        super()
+
+        this._dbConnections = getNamedDBs(databaseUrl || conf['DATABASE_URL'])
+        const availableDatabases = Object.keys(this._dbConnections)
+        this._replicaPoolsConfig = getReplicaPoolsConfig(replicaPools || conf['DATABASE_POOLS'], availableDatabases)
+        this._routingRules = getQueryRoutingRules(routingRules || conf['DATABASE_ROUTING_RULES'], this._replicaPoolsConfig)
+    }
+
+    async _initKnexClients () {
+        const dbNames = Object.keys(this._dbConnections)
+        const connectionResults = await Promise.allSettled(
+            dbNames.map(dbName => initKnexClient({
+                client: 'postgres',
+                pool: { min: 0, max: 3 },
+                connection: this._dbConnections[dbName],
+            }))
+        )
```

Comment:
Any validations before init knex adapter?

Reply:
All configs are validated inside constructor...
Connection is validated inside `initKnexClient`

What else you expect to validate?

Reply:
Ok, I found it. Seems like that's enough